### ISSUE TITLE
Add org.jooq.Meta implementation that is based on DDL interpretation

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#8528]
Similar to #7034, if we had a way to interpret (possibly parsed) DDL statements, then we could expose the resulting meta model through `org.jooq.Meta` and make it available to all client code that uses `org.jooq.Meta`, including the `DDLDatabase` itself.

Such a interpretation would offer itself to allow for extracting different snapshots at different moments of the interpretation, which could evolve into a proper feature.

----

The discussion here https://github.com/jOOQ/jOOQ/issues/8528#issuecomment-530238124 leads to the following list of prerequisites:

- [x] #6412 `DSLContext.ddl()` should distinguish between tables and views
- [x] #7752 Add support for Sequence flags in meta model
- [x] #7977 ~Add an SPI to generate names for unnamed QueryParts~ (postponed for later)
- [x] #9238 Add a new `Settings.parseNameCase`
- [x] #9239 ~Add a new `Settings.parseMetaLookupCaseSensitivity`~ (postponed for later)
- [x] #9633 Add a new `Settings.interpreterMetaLookupCaseSensitivity` 
- [x] #9327 Add `DDLFlag.SEQUENCE` to allow for exporting `CREATE SEQUENCE` statements from `DSLContext.ddl()`
- [x] #9362 A `Settings` or SPI to decide how DDL should be interpreted
- [x] #9364 Add `DDLFlag.INDEX` to allow for exporting `CREATE INDEX` statements from `DSLContext.ddl()`

----

### Subtasks:

Remarks

- All of the statements automatically support `IF EXISTS` or `IF NOT EXISTS` clauses, so they will not be listed explicitly below.
- The same is true for `CASCADE` and `RESTRICT`

Tasks (Statements - list might be incomplete)

- [x] Schema DDL
  - [x] `CREATE SCHEMA`
  - [x] `ALTER SCHEMA`
    - [x] `ALTER SCHEMA .. RENAME TO`
  - [x] `DROP SCHEMA`
    - [x] `DROP SCHEMA .. [ CASCADE | RESTRICT ]`
- [x] Table DDL
  - [x] `CREATE TABLE`
    - [x] `CREATE TEMPORARY TABLE`
  - [x] `DROP TABLE`
    - [x] `DROP TEMPORARY TABLE`
  - [x] `ALTER TABLE`
    - [x] `ALTER TABLE .. RENAME TO`
    - [x] `ALTER TABLE .. RENAME COLUMN .. TO`
    - [x] `ALTER TABLE .. RENAME CONSTRAINT .. TO`
    - [x] `ALTER TABLE .. ADD COLUMN`
    - [x] `ALTER TABLE .. ADD PRIMARY KEY`
    - [x] `ALTER TABLE .. ADD UNIQUE`
    - [x] `ALTER TABLE .. ADD FOREIGN KEY`
    - [x] `ALTER TABLE .. ADD CHECK`
    - [x] `ALTER TABLE .. ALTER COLUMN .. [ SET | DROP ] NOT NULL`
    - [x] `ALTER TABLE .. ALTER COLUMN .. SET DEFAULT` (needs parser support first: #9353)
    - [x] `ALTER TABLE .. ALTER COLUMN .. DROP DEFAULT` (needs API and parser support first: #9354)
    - [x] `ALTER TABLE .. ALTER COLUMN .. SET TYPE`
    - [x] `ALTER TABLE .. DROP COLUMN`
    - [x] `ALTER TABLE .. DROP CONSTRAINT`
    - [x] `ALTER TABLE .. DROP PRIMARY KEY`
    - [x] `ALTER TABLE .. DROP FOREIGN KEY`
  - [x] `COMMENT ON TABLE`
  - [x] `COMMENT ON COLUMN`
  - [x] `TRUNCATE TABLE`
- [x] Index DDL (needs new `DDLFlag` API, first: #9364)
  - [x] `CREATE INDEX`
  - [x] `ALTER INDEX .. RENAME TO`
  - [x] `DROP INDEX`
- [x] View DDL
  - [x] `CREATE VIEW`
  - [x] `ALTER VIEW RENAME TO`
  - [x] `DROP VIEW`
  - [x] `COMMENT ON VIEW`
- [x] Sequence DDL
  - [x] `CREATE SEQUENCE`
  - [x] `ALTER SEQUENCE`
    - [x] `ALTER SEQUENCE .. RENAME TO`
    - [x] `ALTER SEQUENCE .. [ RESTART | RESTART WITH ]` (depends on #7752)
  - [x] `DROP SEQUENCE`
- [x] ~DML~ (postponed: depends on #9581)

Tasks (Other)

- [x] ~Associate `SQLState` to various exceptions~ (might be implemented later)
- [x] Support identities
- [x] ~Which dialects require PKs to be non-null?~ (we don't care about this yet)
- [x] What happens to views that depend upon a table, when the table changes? (we don't care about this yet)
- [x] When renaming objects, make sure all referencing objects follow suit, especially constraints, indexes, etc.
- [x] Support `CASCADE` semantics where supported
  - [x] Explicit support if the flag is given (add API support for `DROP CONSTRAINT`, first: #9380)
  - [x] Default support when:
    - [x] Dropping a column that is part of a constraint -> drop the constraint
    - [x] Dropping a column that is part of an index -> drop the index
- [x] Support `RESTRICT` semantics where supported
  - [x] Explicit support if the flag is given
  - [x] Default support when:
    - [x] Dropping a column that is referenced by a foreign key -> can delete only with explicit `CASCADE`

**Comment by knutwannheden** [comment]
List of open questions (new questions will be added; I would also like to add the answers inline):

- [ ] Should the DDL simulation respect settings like `RenderNameCase` and `RenderQuotedNames`?
    * I suspect yes.
- [ ] Assuming `RenderNameCase` is `AS_IS` and a table's name is unquoted (either explicitly or due to the `RenderQuotedNames` setting), should a table name like `my_table` result in a table with name `MY_TABLE` or `my_table`? I believe the split among the dialects, when it comes to this question, is more or less even.

**Comment by lukaseder** [comment]
> Should the DDL simulation respect settings like RenderNameCase and RenderQuotedNames?

Just as in the `DDLDatabase`, we might use existing settings to *implement* `DDLDatabase` configuration (e.g. `defaultNameCase`), but we should avoid "cleverly reusing" the settings. `RenderNameCase` and `RenderQuotedNames` have very clear meaning in the word `Render`. They define how jOOQ *renders* SQL when sending SQL to some JDBC driver, or when getting `Query.getSQL()`.

But a user who uses `DSLContext::meta` would not expect `RenderNameCase` to have any effect on how DDL is simulated. We will probably need new settings, similar to the ones currently in `DDLDatabase`.

> Assuming RenderNameCase is AS_IS and a table's name is unquoted

I think this question is obsolete once we agree that new settings are needed.

> Should the information schema structure (default catalog / default schema) be fixed or depend on the "parse dialect"?

We will probably have to implement parse dialect specific behaviour in the future. The `DDLDatabase` has a concept of an `unqualifiedSchema`, which we should port to the jOOQ core as a new setting. This covers 90% of all cases, I suspect. What we'll also need:

- `unqualifiedCatalog` (for SQL Server)
- `searchPath` (possibly?) (for PostgreSQL)

Note, we already have a `Settings.parseSearchPath`, which might fit here, for both cases.

**Comment by knutwannheden** [comment]
Another future requirement which I think we should consider in the design: I think the parser should be able to use the interpreter in an incremental fashion, so that it can request a `Meta` representation at any time.

Another twist to this: I think the interpreter will in the future also get a baseline (probably in the form of a `Meta` object) on top of which the DDL is applied.

**Comment by lukaseder** [comment]
> Another future requirement which I think we should consider in the design: I think the parser should be able to use the interpreter in an incremental fashion, so that it can request a Meta representation at any time.

That's not a future requirement :) https://github.com/jOOQ/jOOQ/issues/9091. We'll definitely need this for some emulations, e.g. when we want to emulate MySQL's `DROP PRIMARY KEY` in other RDBMS, which don't have that syntax (i.e. not specifying the PK's name), we can easily look it up from the so-far simulated DDL. We can already look it up from a baseline `Meta` lookup, but it will be even more powerful if we can do it from any step in a DDL script.

> Another twist to this: I think the interpreter will in the future also get a baseline (probably in the form of a Meta object) on top of which the DDL is applied.

Yes, right: https://github.com/jOOQ/jOOQ/issues/9322

Also, I will now rename all the references to "simulation" to "interpretation" and "simulator" to "interpreter". I like your term better than mine.

**Comment by lukaseder** [comment]
I have now refactored the DDL interpretation to work with a new, internal meta model that is fully mutable, and has no functionality, rather than reusing the various `CatalogImpl`, `SchemaImpl`, `TableImpl`, etc. APIs.

Will improve on the simulation by adding more features to it over the next days.

**Comment by lukaseder** [comment]
In the context of all of this, I've just thought we could (finally) start parsing Liquibase XML files and interpret those like any other DDL. Their supported operations are all quite low level, so surely, we can support most of them through jOOQ API and offer the service as a new implementation of `DSLContext.meta(Source...)` as well as in `jOOQ-meta` as a new `LiquibaseDatabase`, similar to the existing `DDLDatabase`.

This way, if people use Liquibase and jOOQ, they can profit from a much tighter integration, with faster code generation, than if they had to first produce the schema / migration in an actual database instance.

There's already an old issue for this.
https://github.com/jOOQ/jOOQ/issues/2286

**Comment by lukaseder** [comment]
New tasks in this area will be implemented in separate issues, and in 3.14 only.
